### PR TITLE
Fix ReturnBinaryOrToEarlyReturnRector mixing up returned nodes of array, hook into StmtsAwareInterface instead to keep next Rector rules updated

### LINKED ISSUE #8044: `System error: "Trying to replace statement (Stmt_Return) with expression (Expr_BooleanNot). Are you missing a Stmt_Expression wrapper?`
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | 0.17.3            |

Not sure which rule it is exactly. But the following snippet causes an internal PHPParser error.

```php
<?php

namespace App;

use Illuminate\Support\Facades\Auth;
use User;

class Foo
{
    public function getUser(?User $user = null): ?User
    {
        return $user ?? (Auth::check() ? Auth::user() : null);
    }


    public function doFoo(
        ?User $user = null
    ): bool {
        $user = $this->getUser($user);

        if (!$user) {
            return false;
        }

        return $user->isFoo() || $user->isBar();
    }

}
```

Error: `System error: "Trying to replace statement (Stmt_Return) with expression (Expr_BooleanNot). Are you missing a Stmt_Expression wrapper?`

## Minimal PHP Code Causing Issue

https://getrector.com/demo/44c449cb-8ecf-420a-b7ef-00018292a101

## Expected Behaviour

No errors.

### PR DESCRIPTION
**Pull request by TomasVotruba**
* fixes https://github.com/rectorphp/rector-src/pull/4474 
* fixes https://github.com/rectorphp/rector/issues/8044

After 3 hours of debugging, I've found out the issue is in 1 node being replaced by many nodes. Espcially `return` with many `ifs`. Lot of rules hook into the `If_` node then and php-parser incorrectly resolved original position. Instead of working with new array, it still operates with original tree. Then the node tree gets messed up as some rules work with one node and another with array of many nodes.

**This is trail of making the rule work with parent node - `StmtsAwareInterface`, to always provide child node with final rules.**

In short: replace one node with array of nodes only on a stmts aware. Better do not change node structure from 1 to many :) 

Let's see how it works, we might want to update few more rules in the future.

### CODE REVIEW

**Comment by TomasVotruba** [comment]
Let's see if this helps with the build :)

**Comment by samsonasik** [comment]
@TomasVotruba thank you 🙏

**Comment by TomasVotruba** [comment]
Thanks!

That was a journey deep into AST world :laughing:
